OP_CODEGEN_GUIDE.md

```diff
@@ -0,0 +1,209 @@
+- For every op, it will generate a {OP}OutputShape method. We need to manually declare and implement this method in a separate file.
+- For every op, it will generate a Lower declaration. We need to manually implement this lowering function in a separate file.
+
+### 3. Implement the missing IR function
+#### torch_xla/csrc/ops/ops_xla_shape_fn.h 
+Declare the {OP}OutputShape:
+```
+xla::Shape AbsOutputShape(const XlaValue& input);
+```
+#### torch_xla/csrc/ops/ops_xla_shape_fn.cpp
+Implement the {OP}OutputShape:
+```
+xla::Shape AbsOutputShape(const XlaValue& input) { return input.xla_shape(); }
+```
+
+`Abs` is an overly simplified example, in a normal case you need to call the BuildXXXOp function again to get the output shape. A slightly better example would be:
+```
+xla::Shape MaximumOutputShape(const XlaValue& input, const XlaValue& other) {
+  auto lower_for_shape_fn =
+      [&](absl::Span<const xla::XlaOp> operands) -> xla::XlaOp {
+    auto promoted = XlaHelpers::Promote(operands[0], operands[1]);
+    return xla::Max(promoted.first, promoted.second);
+  };
+  return InferOutputShape({input.xla_shape(), other.xla_shape()},
+                          lower_for_shape_fn);
+}
+```
+
+Note that you should not start from scratch. Find the Xla::Shape computation logic from the existing op and move it this these two files.
+
+### 4. Implement the lowering function
+#### torch_xla/csrc/ops/ops_lower_fn.cpp
+```
+torch_xla::XlaOpVector Abs::Lower(LoweringContext* loctx) const {
+  xla::XlaOp xla_input = loctx->GetOutputOp(operand(0));
+  return ReturnOp(BuildAbs(xla_input), loctx);
+}
+```
+Note that this function should be directly moved from the existing lowering. Some Ops that were originally implemented in `torch_xla/csrc/ops/ops.cpp` use `GenericOp`. You will need to slightly modify their lowering implementation to fit the implementation provided above.
+
+### 5. Cleanup
+Delete the existing op from aten_xla_type.cpp, tensor.h, tensor_methods.cpp, and ops/…. Note that sometimes you have to keep the tensor_method, because it is being used in tensor_ops like. So, before removing the op, cross reference it with `tensor_ops.cpp`.
+```
+  XLATensor s1 = XLATensor::sub(XLATensor::mul(u2, v3), XLATensor::mul(u3, v2), one);
+```
+Sometimes other IRNode uses the 'IRNode' you migrated. In this case you need to update those IRNode lowering logic as well. In the long term we need to get rid  of these composite IR from our end and provide a lowering function for each op.
+```
+  torch::lazy::NodePtr exp = Pow(Abs(input), norm_exp);
+```
+to
+```
+  torch::lazy::NodePtr exp =
+      Pow(torch::lazy::MakeNode<Abs>(input, std::vector<torch::lazy::Shape>()),
+          norm_exp);
+```
+
+## Run the test and verify the result
+Run the C++ op test or a simple test that only involves the generated ops. As usual, two things to verify are the correctness and the xla counter being incremented correctly.
```

Comment:
Please provide an example of how to run c++ test.

Reply:
Updated